static/src/javascripts/projects/common/modules/commercial/slice-adverts.js

```diff
@@ -1,120 +1,120 @@
+                    adSlice: container.querySelector(sliceSelector)
+                };
+            })
             // pull out ad slices which have at least x containers between them
-            while (index < containers.length) {
-                container    = containers[index];
-                containerId  = bonzo(container).data('id');
-                $adSlice     = $(opts.sliceSelector, container);
-                // don't display ad in the first container on the network fronts
-                isFrontFirst = contains(['uk', 'us', 'au'], config.page.pageId) && index === 0;
-
+            .filter(function (item, index) {
                 if (config.page.showMpuInAllContainers) {
-                    adSlices.push($adSlice.first());
-                    index++;
-                } else {
-                    if ($adSlice.length && !isFrontFirst && (!prefs || prefs[containerId] !== 'closed')) {
-                        adSlices.push($adSlice.first());
-                        index += (containerGap + 1);
-                    } else {
-                        index++;
-                    }
+                    return true;
                 }
-            }
 
-            return Promise.all(chain(adSlices).slice(0, maxAdsToShow).and(map, function ($adSlice, index) {
-                    // When we are inside the AB test we are adding inline1 manually so index needs to start from 2.
-                    var inlineIndexOffset = (config.tests.cmTopBannerPosition) ? 2 : 1;
+                // Slots are inserted every containerGap + 1 slice
+                if (!item.adSlice || index % (containerGap + 1) !== 0) {
+                    return false;
+                }
 
-                    var adName        = 'inline' + (index + inlineIndexOffset),
-                        $mobileAdSlot = bonzo(createAdSlot(adName, 'container-inline'))
-                            .addClass('ad-slot--mobile'),
-                        $tabletAdSlot = bonzo(createAdSlot(adName, 'container-inline'))
-                            .addClass('ad-slot--not-mobile');
+                var containerId  = item.container.getAttribute('data-id');
+                var isContainerClosed = prefs[containerId] === 'closed';
+                var isFrontFirst = isNetworkFront && index === 0;
+                return !(isFrontFirst || isContainerClosed);
+            })
+            // limit to maxAdsToShow
+            .slice(0, maxAdsToShow)
+            // create ad slots for the selected slices
+            .map(function (item, index) {
+                // When we are inside the AB test we are adding inline1 manually so index needs to start from 2.
+                var inlineIndexOffset = (config.tests.cmTopBannerPosition) ? 2 : 1;
+                var adName = 'inline' + (index + inlineIndexOffset);
+                var adSlot = hasFabricAd && index === 0 ?
+                    (fabricAdSlot = createAdSlot('fabric', 'container-inline')) :
```

Comment:
How do we feel about using assignments as expression values? It does make things concise, but once the code grows, does it bury the logic?